Drivers: Add MAX32 SPIXF flash driver

petejohanson-adi opened the pull request:
Adding a flash driver for the MAX32 SPIXF peripheral, supporting attached NOR flash. 

Test run on the APARD32690 kit:

```
west twister --device-testing  --device-serial /dev/tty.usbmodem0001760024391 -p apard32690/max32690/m4 -T tests/drivers/flash/common --device-flash-timeout=30  --west-runner jlink --west-flash -n
Keeping artifacts untouched
INFO    - Using Ninja..
INFO    - Zephyr version: v4.1.0-4226-g0323c0c2054f
INFO    - Using 'zephyr' toolchain.
INFO    - Building initial testsuite list...

Device testing on:

| Platform               | ID   | Serial device                  |
|------------------------|------|--------------------------------|
| apard32690/max32690/m4 |      | /dev/tty.usbmodem0001760024391 |

INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
WARNING - TS_START: already STARTED 'flash_driver':{'count': 1, 'repeat': 0} 19, failed:    0, error:    0
WARNING - TC_START: already STARTED case 'drivers.flash.common.max32_spixf_nor.flash_driver.flash_copy':{'count': 1}
WARNING - TS_START: already STARTED 'flash_driver':{'count': 2, 'repeat': 1}
WARNING - TC_START: already STARTED case 'drivers.flash.common.max32_spixf_nor.flash_driver.flash_copy':{'count': 1}
WARNING - TS_START: already STARTED 'flash_driver':{'count': 3, 'repeat': 2}
WARNING - TC_START: already STARTED case 'drivers.flash.common.max32_spixf_nor.flash_driver.flash_copy':{'count': 2}
WARNING - TS_START: already STARTED 'flash_driver':{'count': 4, 'repeat': 3}
WARNING - TC_START: already STARTED case 'drivers.flash.common.max32_spixf_nor.flash_driver.flash_copy':{'count': 3}
WARNING - TC_START: already STARTED case 'drivers.flash.common.max32_spixf_nor.flash_driver.flash_erase':{'count': 1}
INFO    - Total complete:    4/   4  100%  built (not run):    0, filtered:   19, failed:    0, error:    0
INFO    - 20 test scenarios (20 configurations) selected, 19 configurations filtered (16 by static filter, 3 at runtime).
INFO    - 1 of 1 executed test configurations passed (100.00%), 0 built (not run), 0 failed, 0 errored, with no warnings in 51.45 seconds.
INFO    - 6 of 6 executed test cases passed (100.00%) on 1 out of total 1059 platforms (0.09%).
INFO    - 1 selected test cases not executed: 1 skipped.
INFO    - 1 test configurations executed on platforms, 0 test configurations were only built.

Hardware distribution summary:

| Board                  | ID   |   Counter |   Failures |
|------------------------|------|-----------|------------|
| apard32690/max32690/m4 |      |         1 |          0 |
INFO    - Saving reports...
INFO    - Writing JSON report /Users/petejohanson/source/upstream-zephyr-workspace/zephyr/twister-out/twister.json
INFO    - Writing xunit report /Users/petejohanson/source/upstream-zephyr-workspace/zephyr/twister-out/twister.xml...
INFO    - Writing xunit report /Users/petejohanson/source/upstream-zephyr-workspace/zephyr/twister-out/twister_report.xml...
INFO    - Run completed
```

This does support XIP, but that requires OpenOCD changes that are still working on getting published, in order to flash code relocated there.

petejohanson-adi committed:
drivers: flash: Add MAX32 SPIXF NOR flash driver

Implement support for NOR flash on the SPIXF peripheral found
on MAX32 devices.

Signed-off-by: Pete Johanson <pete.johanson@analog.com>

petejohanson-adi committed:
drivers: flash: Add MAX32 SPIXF NOR test/samples

Update common flash test and jesd216 samples to work on the
APARD32690-SL board which has MX25U6432FM2I02 on-board.

Signed-off-by: Pete Johanson <pete.johanson@analog.com>

petejohanson-adi committed:
drivers: flash: Support XIP for MAX32 SPIXF flash

Properly configure the MAX32 SPIXF peripheral to use the SPIXF controller
for transparent memory mapped reads, and enable the SPIXF main controller
and use it for writes.

Add support for testing XIP support to the nocopy sample, which requires
flashing with OpenOCD with MAX32690 QSPI flash support.

Signed-off-by: Pete Johanson <pete.johanson@analog.com>

MaureenHelm commented:
@de-nordic can you take a look?

de-nordic commented on drivers/flash/flash_max32_spixf_nor.c:
```diff
@@ -0,0 +1,1327 @@
+	/* Run NOR init */
+	ret = flash_max32_spixf_nor_fetch_jesd216_details(dev);
+	if (ret < 0) {
+		return -ENODEV;
```
When return from one function gets replaced with other, the original value should probably be logged.

petejohanson-adi replied:
Logged.

de-nordic reviewed (changes requested):
`size` == 0 processing from _read, _write and _erase to fix.
Loops without timeout to fix.
Other are just comments.

petejohanson-adi committed:
drivers: flash: Enable static/runtime SFDP data for SPIXF driver

Add a new flash for toggling runtime SFDP data fetching, and use static
devicetree properties by default. Clean up various minor items from review.

Signed-off-by: Pete Johanson <pete.johanson@analog.com>

petejohanson-adi commented on drivers/flash/flash_max32_spixf_nor.c:
```diff
@@ -0,0 +1,1327 @@
+	/* erase non-zero size */
+	if (size == 0) {
+		return 0;
+	}
```
Fixed.

petejohanson-adi commented on drivers/flash/flash_max32_spixf_nor.c:
```diff
@@ -0,0 +1,1327 @@
+
+	uint32_t erase_size = BIT(exponent);
+
+	/* We need layout page size to be compatible with erase size */
+	if ((layout_page_size % erase_size) != 0) {
+		LOG_DBG("layout page %u not compatible with erase size %u", layout_page_size,
+			erase_size);
+		LOG_DBG("erase size will be used as layout page size");
+		layout_page_size = erase_size;
+	}
+
+	/* Warn but accept layout page sizes that leave inaccessible
+	 * space.
+	 */
+	if ((flash_size % layout_page_size) != 0) {
+		LOG_INF("layout page %u wastes space with device size %zu", layout_page_size,
+			flash_size);
+	}
+
+	data->layout.pages_size = layout_page_size;
+	data->layout.pages_count = flash_size / layout_page_size;
+	LOG_DBG("layout %u x %u By pages", data->layout.pages_count, data->layout.pages_size);
```
I've switched this modeled after other drivers to either support runtime SDPF or leverage fixed values from devicetree (the default), and added another entry to the test case so both get exercised in testing.

petejohanson-adi commented on drivers/flash/flash_max32_spixf_nor.c:
```diff
@@ -0,0 +1,1327 @@
+
+	do {
+		ret = qspi_read_status_register(dev, 1U, &reg);
+	} while (!ret && !(reg & SPI_NOR_WEL_BIT));
```
Added a fixed number of retries before bailing.

petejohanson-adi commented on drivers/flash/flash_max32_spixf_nor.c:
```diff
@@ -0,0 +1,1327 @@
+	return ret;
+}
+
+static int spi_nor_process_bfp(const struct device *dev, const struct jesd216_param_header *php,
```
See my other comment. This function is still used to load things from the devicetree, unless you enable the new runtime Kconfig option.

petejohanson-adi commented:
@de-nordic Looking forward to a review when you have a moment. Thanks!

kartben reviewed (dismissed):
some minor typos but the one in the SPDX license header is I think worth the request for change - hope you won't mind!

petejohanson-adi commented:
> some minor typos but the one in the SPDX license header is I think worth the request for change - hope you won't mind!

Not at all, thanks for the review! Addressed all your comments, and rebased on latest `main` while at it.

MaureenHelm commented:
> Note: The test failures are unrelated, and should be fixed once #91600 is merged.

Reran the twister job after #91600 was merged and now it's green.

MaureenHelm commented:
@de-nordic can you take another look?

de-nordic reviewed (dismissed):
Looks OK. Some comments left regarding logging, can be addressed later.

kartben commented on dts/bindings/flash_controller/adi,max32-spixf-nor.yaml:
```diff
@@ -0,0 +1,73 @@
+            compatible = "adi,max32-spixf-nor";
+            reg = <0x0 DT_SIZE_M(8)>; /* 64 Mbits */
+            qspi-max-frequency = <80000000>;
+            jedec-id = [25 37 c2];
```
@de-nordic Yeah, I know, I did check the datasheet :) But I was asking in case for some reason the bytes were to be indicated in a different order in the DTS, but based on your answer and other similar flash chips used in the tree that also start with c2 xx yy, it does look like it needs fixing :)

kartben replied:
Btw @de-nordic, shouldn't this be caught in the test suite?

de-nordic replied:
> Btw @de-nordic, shouldn't this be caught in the test suite?

Other drivers, like spi_nor.c, use JEDEC ID to figure out whether they work with what they expect by DTS, so basically that driver would not work if JEDEC ID would not match. But they know what to expect (*).

We have `flash_read_jedec_id`, so we probably should have it tested, with common Flash API tests,  or doing verification against what DTS states, the problem is that the functions unusable in user code... completely. And that is because of the specification, https://www.jedec.org/standards-documents/docs/jep-106ab, that vendors ignore the specification it, so you have to basically follow chip datasheet instead, and overall function design in the end.

So what is wrong with the `flash_read_jedec_id`:
 - It requires buffer, that nobody knows what size should be, maybe driver knows, but no way to ask it.
 - you may need to allocate more than 17 bytes for ID, but still you do not know that.
 - It is pointless to ask for ID, as vendors often ignore requirement for "continuation codes" (7f), so you can not tell memories apart.

Example:
 - Some Macronix IC ID: c2 xx xx
 - Some Valkyrie IC ID: 7f 7f 7f 7f 7f 7f 7f 7f 7f 7f 7f 7f 7f 7f c2 yy yy

If RDID only returns three last bytes, you can only rely on vendor not matching the last bytes, so you can barely tell the chips apart. And vendors can put more bytes then three.
This is in our tree:
https://github.com/zephyrproject-rtos/zephyr/blob/7900f8d4ee7b318c0813c9a5629d000cdae7b06c/boards/nxp/s32z2xxdc2/s32z2xxdc2_s32z270.dtsi#L105
(aside from the  ID being completely wrong, because 00 is not a valid  ID, and if it is backwards then 90 is not valid either).

That is why (*) even driver has to know what to expect and still may be wrong.

petejohanson-adi commented on drivers/flash/flash_max32_spixf_nor.c:
```diff
@@ -0,0 +1,1377 @@
+				payload[0] = bet->cmd;
+				ret = qspi_send_req(dev, &req);
+				if (ret < 0) {
+					LOG_ERR("Failet to do a erase (%d)", ret);
```
Fixed.

petejohanson-adi commented on drivers/flash/flash_max32_spixf_nor.c:
```diff
@@ -0,0 +1,1377 @@
+	int rc;
+
+	if (flash_size != dev_cfg->flash_size) {
+		LOG_ERR("Unexpected flash size: %u", flash_size);
```
Added a return here. Thanks.

petejohanson-adi commented on drivers/flash/flash_max32_spixf_nor.c:
```diff
@@ -0,0 +1,1377 @@
+	{.bus = DT_CLOCKS_CELL_BY_IDX(node_id, idx, offset),                                       \
+	 .bit = DT_CLOCKS_CELL_BY_IDX(node_id, idx, bit)}
+
+/* clang-format off */
```
Yeah, annoying indeed.

petejohanson-adi commented on drivers/flash/flash_max32_spixf_nor.c:
```diff
@@ -0,0 +1,1377 @@
+		ret = qspi_read_status_register(dev, 1U, &reg);
+	} while (!ret && !(reg & SPI_NOR_WEL_BIT) && (retries++) < WRITE_ENABLE_MAX_RETRIES);
+
+	return ret;
```
Added a condition for this. Thanks.

petejohanson-adi commented on drivers/flash/flash_max32_spixf_nor.c:
```diff
@@ -0,0 +1,1377 @@
+	return qspi_send_req(dev, &req);
+}
+
+#define WRITE_ENABLE_MAX_RETRIES 4
```
Moved to the top of the unit.

petejohanson-adi commented:
Thanks @kartben and @de-nordic for your reviews. Addressed all the comments, hopefully we can get the approvals and merge this before the 4.2 cutoff.

de-nordic commented on boards/adi/apard32690/apard32690_max32690_m4.dts:
```diff
@@ -308,3 +308,26 @@ pmod_spi: &spi4 {
+		compatible = "adi,max32-spixf-nor";
+		reg = <0x08000000 DT_SIZE_M(8)>; /* 64 Mbits */
+		qspi-max-frequency = <60000000>;
+		jedec-id = [25 37 c2];
```
https://github.com/zephyrproject-rtos/zephyr/pull/90175#discussion_r2171218550 here

kartben commented:
@petejohanson-adi as a general note it's usually easier on reviewers to not "unnecessarily" rebase your PR when you make minor amends and you have nothing you need to pick from main, this way the "compare" UI in gihub is clean and showing only the diff with the previous PR and making it super easy to re-review :)
